File: src/Lumineux.Application/Contracts/Auth/AuthDtos.cs
namespace Lumineux.Application.Contracts.Auth;

/// <summary>Requête de connexion (FR-001).</summary>
public sealed record LoginRequest(string Reference, string Password);

/// <summary>Requête d'activation / première connexion (FR-007).</summary>
public sealed record ActivateAccountRequest(string Reference, string TemporaryPassword, string NewPassword);

/// <summary>Requête de changement de mot de passe pour un utilisateur connecté (FR-009).</summary>
public sealed record ChangePasswordRequest(string CurrentPassword, string NewPassword);

/// <summary>Réponse contenant le jeton d'accès. Aucun mot de passe n'est exposé.</summary>
public sealed record TokenResponse(string AccessToken, string TokenType, DateTime ExpiresAt);

/// <summary>Requête de demande de réinitialisation de mot de passe (feature 006, FR-001).</summary>
public sealed record ForgotPasswordRequest(string Reference);

/// <summary>Requête de réinitialisation avec le jeton reçu par email (feature 006, FR-005).</summary>
public sealed record ResetPasswordRequest(string Token, string NewPassword);

/// <summary>Réponse générique anti-énumération (feature 006, FR-002). Aucun détail sur le compte.</summary>
public sealed record GenericMessageResponse(string Message);

/// <summary>
/// Profil de session de l'utilisateur courant (feature 007, FR-004/005). Dérivé du jeton : identité
/// minimale + droits effectifs de la session. Aucune donnée secrète n'est exposée (FR-007).
/// </summary>
public sealed record CurrentUserResponse(int MemberId, string DisplayName, IReadOnlyList<string> Permissions);

File: mobile/lib/features/auth/data/auth_dtos.dart
/// DTO **miroir** des contrats serveur (`Lumineux.Application.Contracts.Auth`).
/// Sérialisation manuelle (petit ensemble). Aucun mot de passe n'est stocké
/// ni journalisé au-delà de la durée de l'appel.
library;

class LoginRequest {
  const LoginRequest({required this.reference, required this.password});

  final String reference;
  final String password;

  Map<String, dynamic> toJson() => {
        'reference': reference,
        'password': password,
      };
}

class ActivateRequest {
  const ActivateRequest({
    required this.reference,
    required this.temporaryPassword,
    required this.newPassword,
  });

  final String reference;
  final String temporaryPassword;
  final String newPassword;

  Map<String, dynamic> toJson() => {
        'reference': reference,
        'temporaryPassword': temporaryPassword,
        'newPassword': newPassword,
      };
}

class ForgotPasswordRequest {
  const ForgotPasswordRequest({required this.reference});

  final String reference;

  Map<String, dynamic> toJson() => {'reference': reference};
}

class ResetPasswordRequest {
  const ResetPasswordRequest({required this.token, required this.newPassword});

  final String token;
  final String newPassword;

  Map<String, dynamic> toJson() => {
        'token': token,
        'newPassword': newPassword,
      };
}

class ChangePasswordRequest {
  const ChangePasswordRequest({
    required this.currentPassword,
    required this.newPassword,
  });

  final String currentPassword;
  final String newPassword;

  Map<String, dynamic> toJson() => {
        'currentPassword': currentPassword,
        'newPassword': newPassword,
      };
}

class TokenResponse {
  const TokenResponse({
    required this.accessToken,
    required this.tokenType,
    required this.expiresAt,
  });

  final String accessToken;
  final String tokenType;
  final DateTime expiresAt;

  factory TokenResponse.fromJson(Map<String, dynamic> json) => TokenResponse(
        accessToken: json['accessToken'] as String,
        tokenType: (json['tokenType'] as String?) ?? 'Bearer',
        expiresAt: DateTime.parse(json['expiresAt'] as String),
      );
}

class CurrentUser {
  const CurrentUser({
    required this.memberId,
    required this.displayName,
    required this.permissions,
  });

  final String memberId;
  final String displayName;
  final List<String> permissions;

  factory CurrentUser.fromJson(Map<String, dynamic> json) => CurrentUser(
        memberId: json['memberId'].toString(),
        displayName: (json['displayName'] as String?) ?? '',
        permissions: (json['permissions'] as List<dynamic>? ?? const [])
            .map((e) => e.toString())
            .toList(growable: false),
      );
}
